Content.Client/Visualizer/GenericEnumVisualizer.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using JetBrains.Annotations;
+using Robust.Client.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Log;
+using Robust.Shared.Reflection;
+using Robust.Shared.Serialization;
+using Robust.Shared.Serialization.Manager.Attributes;
+
+namespace Content.Client.Visualizer
+{
+    [UsedImplicitly]
+    public class GenericEnumVisualizer : AppearanceVisualizer, ISerializationHooks
+    {
+        public Enum Key { get; set; } = default!;
+
+        public Dictionary<object, string> States { get; set; } = default!;
+
+        [DataField("layer")]
+        public int Layer { get; set; } = 0;
+
+        [DataField("key", readOnly: true, required: true)]
+        private string _keyRaw = default!;
+
+        [DataField("states", readOnly: true, required: true)]
+        private Dictionary<string, string> _statesRaw { get; set; } = default!;
+
+        void ISerializationHooks.AfterDeserialization()
+        {
+            var reflectionManager = IoCManager.Resolve<IReflectionManager>();
+
+            object ResolveRef(string raw)
+            {
+                if (reflectionManager.TryParseEnumReference(raw, out var @enum))
+                {
+                    return @enum;
+                }
+                else
+                {
+                    Logger.WarningS("c.c.v.genum", $"Unable to convert enum reference: {raw}");
+                }
+
+                return raw;
+            }
+
+            // It's important that this conversion be done here so that it may "fail-fast".
+            Key = (Enum) ResolveRef(_keyRaw);
+            States = _statesRaw.ToDictionary(kvp => ResolveRef(kvp.Key), kvp => kvp.Value);
+        }
```

Comment:
god i just want a way to avoid having to do this. i already hate it in userinterfacecomponent

Reply:
funny you mention that component...
that was more or less what acted as the example